sdks/python/apache_beam/io/hdfilesystem.py

```diff
@@ -0,0 +1,190 @@
+
+  def mkdirs(self, path):
+    if self.exists(path):
+      raise IOError('%s: path already exists' % path)
+    self._hdfs.makedirs(path)
+
+  def match(self, patterns, limits=None):
+    if limits is None:
+      limits = [None] * len(patterns)
+    else:
+      err_msg = 'Patterns and limits should be equal in length'
+      assert len(patterns) == len(limits), err_msg
+
+    def _match(pattern, limit):
+      """Find all matching paths to the pattern provided."""
+      fileinfos = self._hdfs.ls(pattern, detail=True)[:limit]
+      metadata_list = [FileMetadata(fileinfo['name'], fileinfo['size'])
+                       for fileinfo in fileinfos]
+      return MatchResult(pattern, metadata_list)
+
+    exceptions = {}
+    result = []
+    for pattern, limit in zip(patterns, limits):
+      try:
+        result.append(_match(pattern, limit))
+      except Exception as e:  # pylint: disable=broad-except
+        exceptions[pattern] = e
+
+    if exceptions:
+      raise BeamIOError('Match operation failed', exceptions)
+    return result
+
+  def _open(self, path, mode, mime_type, compression_type):
+    if mime_type != 'application/octet-stream':
+      raise BeamIOError('Unsupported mime_type: %s', mime_type)
+    if compression_type == CompressionTypes.AUTO:
+      compression_type = CompressionTypes.detect_compression_type(path)
+    res = self._hdfs.open(path, mode)
+    if compression_type != CompressionTypes.UNCOMPRESSED:
+      res = CompressedFile(res)
+    return res
+
+  def create(self, path, mime_type='application/octet-stream',
+             compression_type=CompressionTypes.AUTO):
+    return self._open(path, 'wb', mime_type, compression_type)
+
+  def open(self, path, mime_type='application/octet-stream',
+           compression_type=CompressionTypes.AUTO):
+    return self._open(path, 'rb', mime_type, compression_type)
+
+  def copy(self, source_file_names, destination_file_names):
+    """Implements the FileSystem interface.
+
+    Will overwrite files and directories in destination_file_names.
+    """
+    err_msg = ("source_file_names and destination_file_names should "
+               "be equal in length")
+    assert len(source_file_names) == len(destination_file_names), err_msg
+
+    def _copy_file(source, destination):
+      with self.open(source) as f1:
+        with self.create(destination) as f2:
+          while True:
+            buf = f1.read(COPY_BUFFER_SIZE)
+            if not buf:
+              break
+            f2.write(buf)
+
+    def _copy_path(source, destination):
+      """Recursively copy the file tree from the source to the destination."""
+      if not self._hdfs.isdir(source):
+        _copy_file(source, destination)
+        return
+
+      for path, dirs, files in self._hdfs.walk(source):
+        for dir in dirs:
+          new_dir = self.join(destination, dir)
+          if not self.exists(new_dir):
+            self.mkdirs(new_dir)
+
+        relpath = posixpath.relpath(path, source)
```

Comment:
<!--thread_id:cc_156305015_t; commit:3c708ac10af4457fe46e07503f0b72930832db07; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Why would 'path' be different from 'source' ?

<!--section:body-->
Say you're copying from /a to /d.
The tree under /a looks like:
/a - dir
/a/b - dir
/a/b/c - file

The result under /d should look like:
/d - dir
/d/b - dir
/d/b/c - file

relpath is what is appended after "/d" ("b" in this case). Otherwise you'd get a file like /d/a/b/c.